src/main/java/org/snomed/snowstorm/core/data/services/CodeSystemVersionService.java

```diff
@@ -0,0 +1,184 @@
+package org.snomed.snowstorm.core.data.services;
+
+import io.kaicode.elasticvc.api.BranchService;
+import io.kaicode.elasticvc.domain.Branch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.snomed.snowstorm.core.data.domain.CodeSystem;
+import org.snomed.snowstorm.core.data.domain.CodeSystemVersion;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.elasticsearch.core.ElasticsearchOperations;
+import org.springframework.data.elasticsearch.core.SearchHit;
+import org.springframework.data.elasticsearch.core.query.NativeSearchQueryBuilder;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.index.query.QueryBuilders.termQuery;
+
+@Component
+public class CodeSystemVersionService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CodeSystemVersionService.class);
+    private static final String SNOMEDCT = "SNOMEDCT";
+
+    /*
+     * K => CodeSystemVersion.toString()
+     * V => CodeSystemVersion.getDependantVersionEffectiveTime()
+     * */
+    private final Map<String, Integer> dependantVersionCache = new HashMap<>();
+
+    @Autowired
+    private CodeSystemService codeSystemService;
+
+    @Autowired
+    private BranchService branchService;
+
+    @Autowired
+    private ElasticsearchOperations elasticsearchOperations;
+
+    /**
+     * Populate cache with the version of SCT the latest version
+     * of the CodeSystem depends on.
+     *
+     * @param codeSystems The CodeSystems to process.
+     */
+    public void initDependantVersionCache(List<CodeSystem> codeSystems) {
+        for (CodeSystem codeSystem : codeSystems) {
+            CodeSystemVersion latestVersion = codeSystem.getLatestVersion();
+            if (latestVersion == null || SNOMEDCT.equals(latestVersion.getShortName())) {
+                continue;
+            }
+            this.doGetDependantVersionForCodeSystemVersion(latestVersion)
+                .ifPresent(dependantVersion -> dependantVersionCache.putIfAbsent(latestVersion.toString(), dependantVersion));
+        }
+    }
+
+    /**
+     * Clear cache.
+     */
+    public void clearCache() {
+        LOGGER.info("Clearing cache 'dependantVersionCache'.");
+        this.dependantVersionCache.clear();
+    }
+
+    /**
+     * Get the version of SCT the CodeSystemVersion depends on.
+     *
+     * @param codeSystemVersion The CodeSystemVersion to process.
+     * @return The version of SCT the CodeSystemVersion depends on.
+     */
+    public Optional<Integer> getDependantVersionForCodeSystemVersion(CodeSystemVersion codeSystemVersion) {
```

Comment:
This method will find what version of SCT the given `CodeSystemVersion` depends on. The method will first try reading from the cache. If the `CodeSystemVersion` is not present in the cache, it will try work out the `dependantVersionEffectiveTime` value from querying Elasticsearch.